### DISCUSSION TITLE
There should be compatibility with local and open source programs

### DISCUSSION THREAD

**Discussion by Pythone** [ThioJoe/Auto-Synced-Translated-Dubs#59] [Ideas]
# Why?
The reason why is to stop relying on these cloud servers entirely. It would also reduce the cost for API credits(local does not require credits, and some open source APIs are self-hostable with changeable servers that it would help).
# examples of These types of projects
## Translation
[LibreTranslate](https://libretranslate.com/) is an open-source self-hostable API. Some servers(including the official one) require credits, while others do not.
[TranslateLocally](https://translatelocally.com/) and [Firefox Translations](https://blog.mozilla.org/en/mozilla/local-translation-add-on-project-bergamot/)  are local tools that are similar to each other and perform a little bit better.
## Text To Speech
[Coqui.ai's open source engine](https://github.com/coqui-ai/tts) works similar to the cloud TTS engines you use. It is a self-hostable API that you could host locally. There are more accurate TTS engines out there(I listed this one to give an example).
# Implementation:
## Translation
For LibreTranslate, You can use [LibreTranslate-py](https://github.com/argosopentech/LibreTranslate-py) for the Python API.
For TranslateLocally, You could probably run the commands from the website in Python.
## TTS
For Coqui.ai, you could use the [examples provided](https://tts.readthedocs.io/en/latest/inference.html)

**Comment by ghost** [comment]
+1 We should not rely on large corporations, and even if you do not hate large corporations, this would allow the script to function without the need for third-party APIs.

**Comment by pankaj-yadav0507** [comment]
Nice suggestion. But comparing both LibreTranslate-py with Google translate is a bit of overstatement. I love open source and I know it's limitations also. But hey you get good amount of free conversations in google translate also. Now where LibreTranslate-py really lacks is no. of languages available. It's really limited for a production use case. On top of it lacks many features like there is no equivalent to `<span translate="no"></span>` in it. 

Also, 
Comparing a translated sample from both, Google translate have a huge advantage.

Sample: 
> LibreTranslate is awesome!

Results (Hindi):
> LibreTranslate -> LibreTranslate भयानक है!
> Google translate -> लिब्रेट्रांसलेट बहुत बढ़िया है! 

Any Hindi speaking person will tell you how wrong LibreTranslate's conversion is.

Haven't tested Coqui.ai so can't comment much on it.  But the base thing is it doesn't work on windows that's a big disappointment even before starting.

**PS: I am not a google translate fanboy just putting some insights since we are using it on production level.**

**Reply by Pythone** [reply]
Thank you for the correction

**Comment by qgustavor** [comment]
I tried LibreTranslate in the past ([I even contributed a little](https://github.com/LibreTranslate/LibreTranslate/pull/182) to the project in the past) and I want to add something: the idea of LibreTranslate is good, but the translation quality is often not great. While it's open-source, I'm starting to think that their models are not going to improve a lot in future since they started to promote the paid training option [in the language requests discussion](https://github.com/argosopentech/argos-translate/discussions/91). That's just a theory of mine and it might be wrong (well, I hope it is wrong), but that's a thing to consider about that project.

I switched to using Helsinki-NLP's models available on Hugging Face. Some models such [this one](https://huggingface.co/Helsinki-NLP/opus-mt-tc-big-itc-itc) allow translating without English being an intermediate language, which was amazing for me since almost every translation service and model uses English as a pivot language. For translating English videos to other languages that's fine, but I was translating Italian subtitles to Portuguese and a lot of context (mostly gender information because many words in Italian and Portuguese are gendered but in English they are not) was being lost during the intermediate English translation.

There are [many translation models](https://huggingface.co/models?pipeline_tag=translation&sort=downloads) in Hugging Face, maybe some performs better than LibreTranslate in tasks other than [non-English language] to [other non-English language]. I don't know Hindi at all, but at least I see that [this translation](https://huggingface.co/Helsinki-NLP/opus-mt-en-hi?text=LibreTranslate+is+awesome%21) looks a bit more similar to the one from Google Translate than from the one from LibreTranslate. Is it better? I don't know, again, I don't know nothing about Hindi, sorry.

Sometimes (well, in some rare cases to be fair) Hugging Face models can be better than Google: from the languages I know something I'm certain that [this translation](https://huggingface.co/Helsinki-NLP/opus-mt-ja-pt?text=%3E%3Epor%3C%3C+%E3%83%84%E3%83%AB) is better than [the one from Google Translate](https://translate.google.com/?sl=auto&tl=pt&text=%E3%83%84%E3%83%AB%E3%81%AF%E5%A4%A7%E3%81%8D%E3%81%AA%E9%B3%A5%E3%81%A7%E3%81%99%E3%80%82&op=translate): because Google uses English as an intermediate language it first translates "ツル" ([the crane bird](https://www.google.com/search?tbm=isch&q=%E3%83%84%E3%83%AB&tbs=imgo:1)) into "crane" then translates it into Portuguese, but translates "crane" as being the machine (guindaste), not the bird (grou).

**Reply by Pythone** [reply]
Thanks for telling me

**Comment by Pythone** [comment]
> I tried LibreTranslate in the past ([I even contributed a little](https://github.com/LibreTranslate/LibreTranslate/pull/182) to the project in the past) and I want to add something: the idea of LibreTranslate is good, but the translation quality is often not great. While it's open-source, I'm starting to think that their models are not going to improve a lot in future since they started to promote the paid training option [in the language requests discussion](https://github.com/argosopentech/argos-translate/discussions/91). That's just a theory of mine and it might be wrong (well, I hope it is wrong), but that's a thing to consider about that project.
> 
> I switched to using Helsinki-NLP's models available on Hugging Face. Some models such [this one](https://huggingface.co/Helsinki-NLP/opus-mt-tc-big-itc-itc) allow translating without English being an intermediate language, which was amazing for me since almost every translation service and model uses English as a pivot language. For translating English videos to other languages that's fine, but I was translating Italian subtitles to Portuguese and a lot of context (mostly gender information because many words in Italian and Portuguese are gendered but in English they are not) was being lost during the intermediate English translation.
> 
> There are [many translation models](https://huggingface.co/models?pipeline_tag=translation&sort=downloads) in Hugging Face, maybe some performs better than LibreTranslate in tasks other than [non-English language] to [other non-English language]. I don't know Hindi at all, but at least I see that [this translation](https://huggingface.co/Helsinki-NLP/opus-mt-en-hi?text=LibreTranslate+is+awesome%21) looks a bit more similar to the one from Google Translate than from the one from LibreTranslate. Is it better? I don't know, again, I don't know nothing about Hindi, sorry.
> 
> Sometimes (well, in some rare cases to be fair) Hugging Face models can be better than Google: from the languages I know something I'm certain that [this translation](https://huggingface.co/Helsinki-NLP/opus-mt-ja-pt?text=%3E%3Epor%3C%3C+%E3%83%84%E3%83%AB) is better than [the one from Google Translate](https://translate.google.com/?sl=auto&tl=pt&text=%E3%83%84%E3%83%AB%E3%81%AF%E5%A4%A7%E3%81%8D%E3%81%AA%E9%B3%A5%E3%81%A7%E3%81%99%E3%80%82&op=translate): because Google uses English as an intermediate language it first translates "ツル" ([the crane bird](https://www.google.com/search?tbm=isch&q=%E3%83%84%E3%83%AB&tbs=imgo:1)) into "crane" then translates it into Portuguese, but translates "crane" as being the machine (guindaste), not the bird (grou).

thank you

**Comment by Pythone** [comment]
> Nice suggestion. But comparing both LibreTranslate-py with Google translate is a bit of overstatement. I love open source and I know it's limitations also. But hey you get good amount of free conversations in google translate also. Now where LibreTranslate-py really lacks is no. of languages available. It's really limited for a production use case. On top of it lacks many features like there is no equivalent to `<span translate="no"></span>` in it.
> 
> Also, Comparing a translated sample from both, Google translate have a huge advantage.
> 
> Sample:
> 
> > LibreTranslate is awesome!
> 
> Results (Hindi):
> 
> > LibreTranslate -> LibreTranslate भयानक है!
> > Google translate -> लिब्रेट्रांसलेट बहुत बढ़िया है!
> 
> Any Hindi speaking person will tell you how wrong LibreTranslate's conversion is.
> 
> Haven't tested Coqui.ai so can't comment much on it. But the base thing is it doesn't work on windows that's a big disappointment even before starting.
> 
> **PS: I am not a google translate fanboy just putting some insights since we are using it on production level.**

I am taking your suggestion in. In fact, we could do different models with TranslateLocally. There could be similar programs to Coqui.ai that do work on Windows, otherwise there are bridges.

**Comment by ThioJoe** [comment]
Moving this to a discussion instead

**Reply by Pythone** [reply]
Thanks Thio. The main reason I want this to happen is so that we do not rely on too many services(most require credits) and the only external services needed are: github(for code) and youtube(for the video)
